xarray/backends/rasterio_.py

```diff
@@ -0,0 +1,116 @@
+from collections import OrderedDict
+import numpy as np
+
+try:
+    import rasterio
+except ImportError:
+    rasterio = False
+
+from .. import DataArray
+from ..core.utils import NDArrayMixin, is_scalar
+from ..core import indexing
+
+
+_ERROR_MSG = ('The kind of indexing operation you are trying to do is not '
+              'valid on rasterio files. Try to load your data with ds.load()'
+              'first.')
+
+
+class RasterioArrayWrapper(NDArrayMixin):
```

Comment:
You need to set an `array` property if you use `NDArrayMixin`.

In this case, I would just use the base classes `NdimSizeLenMixin` and `DunderArrayMixin`.